**Market Basket Analysis**

**Market Basket Analysis**, or MBA, is a subset of affinity analysis and has been used in the retail sector for many years. It provides a computational method for identifying common associations between items - usually products - from which strategy can be formed.


The dataset has 20507 entries, over 9000 transactions, and 4 columns.

**Load the packages**

In [ ]:
import pandas as pd

**Load the data**

Here, we’re going to load up some real [transactional data from a bakery business](https://www.kaggle.com/datasets/mittalvasu95/the-bread-basket), calculate the product association rules, and look at how we could use the output to help us create a strategy to increase sales.

In [ ]:
dataset = pd.read_csv('retail_bakery_transactions.csv')

In [ ]:
# Change the data frame's column names to lower case
dataset.columns = dataset.columns.str.lower()

In [ ]:
# Transform every single string inside of the data frame to lower case
dataset = dataset.applymap(lambda s: s.lower() if type(s) == str else s)

In [ ]:
dataset.head()

transaction           item         date_time period_day weekday_weekend
0            1          bread  30-10-2016 09:58    morning         weekend
1            2   scandinavian  30-10-2016 10:05    morning         weekend
2            2   scandinavian  30-10-2016 10:05    morning         weekend
3            3  hot chocolate  30-10-2016 10:07    morning         weekend
4            3            jam  30-10-2016 10:07    morning         weekend

In [ ]:

dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20507 entries, 0 to 20506
Data columns (total 5 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   transaction      20507 non-null  int64 
 1   item             20507 non-null  object
 2   date_time        20507 non-null  object
 3   period_day       20507 non-null  object
 4   weekday_weekend  20507 non-null  object
dtypes: int64(1), object(4)
memory usage: 801.2+ KB


**Check for missing values**

Before moving on, we will check to see if there are any null values to impute. However, the data were all fine, so there was nothing to do.

In [ ]:
dataset.isnull().sum()

transaction        0
item               0
date_time          0
period_day         0
weekday_weekend    0
dtype: int64

**Getting the list of transactions**

Once we have read the dataset, we need to get the list of items in each transaction. This list will work as a training set from where we can generate the list of association rules.

In [ ]:
# Convert the data frame to list
transaction_list = dataset.groupby(['transaction', 'date_time'])['item'].apply(lambda x: list(x))

In [ ]:
transaction_list.head()

transaction  date_time       
1            30-10-2016 09:58                          [bread]
2            30-10-2016 10:05     [scandinavian, scandinavian]
3            30-10-2016 10:07    [hot chocolate, jam, cookies]
4            30-10-2016 10:08                         [muffin]
5            30-10-2016 10:13          [coffee, pastry, bread]
Name: item, dtype: object

In [ ]:
# Converting the data frame into a list of lists
df = transaction_list.values.tolist()

In [ ]:
df[:5]

[['bread'],
 ['scandinavian', 'scandinavian'],
 ['hot chocolate', 'jam', 'cookies'],
 ['muffin'],
 ['coffee', 'pastry', 'bread']]

**One-hot encoding transaction data**

In [ ]:
from mlxtend.preprocessing import TransactionEncoder

In [ ]:
# Instantiate TransactionEncoder and identify unique items
encoder = TransactionEncoder().fit(df)

In [ ]:
# One-hot encode transactions
onehot = encoder.transform(df)

In [ ]:
# Convert one-hot encoded data to data frame
transf_df = pd.DataFrame(onehot, columns = encoder.columns_)

In [ ]:
transf_df.head()

adjustment  afternoon with the baker  alfajores  argentina night  art tray  \
0       False                     False      False            False     False   
1       False                     False      False            False     False   
2       False                     False      False            False     False   
3       False                     False      False            False     False   
4       False                     False      False            False     False   

   bacon  baguette  bakewell  bare popcorn  basket  ...  the bart  the nomad  \
0  False     False     False         False   False  ...     False      False   
1  False     False     False         False   False  ...     False      False   
2  False     False     False         False   False  ...     False      False   
3  False     False     False         False   False  ...     False      False   
4  False     False     False         False   False  ...     False      False   

   tiffin  toast  truffles  tshirt  valentine's card  vegan feast  \
0   False  False     False   False             False        False   
1   False  False     False   False             False        False   
2   False  False     False   False             False        False   
3   False  False     False   False             False        False   
4   False  False     False   False             False        False   

   vegan mincepie  victorian sponge  
0           False             False  
1           False             False  
2           False             False  
3           False             False  
4           False             False  

[5 rows x 94 columns]

**Definition**

Association rules analysis is a technique to uncover how items are associated to each other. There are three common ways to measure association.

**Measure 1: Support.** This says how popular an itemset is, as measured by the proportion of transactions in which an itemset appears.

**Measure 2: Confidence.** This says how likely item Y is purchased when item X is purchased, expressed as {X → Y}.

**Measure 3: Lift.** This says how likely item Y is purchased when item X is purchased, while controlling for how popular item Y is. When lift > 1 then the rule is better at predicting the result than guessing. When lift < 1, the rule is doing worse than informed guessing.

**Run the FP-growth algorithm**

We will generate the association rules using the FP-growth algorithm.


In [ ]:
from mlxtend.frequent_patterns import fpgrowth

In [ ]:
frequent_itemsets = fpgrowth(transf_df, min_support = 0.05, use_colnames = True)

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


In [ ]:
frequent_itemsets.sort_values('support', ascending = False)

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


support         itemsets
3   0.478394         (coffee)
0   0.327205          (bread)
6   0.142631            (tea)
7   0.103856           (cake)
9   0.090016  (coffee, bread)
4   0.086107         (pastry)
8   0.071844       (sandwich)
5   0.061807      (medialuna)
1   0.058320  (hot chocolate)
10  0.054728   (coffee, cake)
2   0.054411        (cookies)

**Examining the frequent itemsets**

In [ ]:
frequent_itemsets['length'] = frequent_itemsets['itemsets'].apply(lambda x: len(x))
frequent_itemsets

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


support         itemsets  length
0   0.327205          (bread)       1
1   0.058320  (hot chocolate)       1
2   0.054411        (cookies)       1
3   0.478394         (coffee)       1
4   0.086107         (pastry)       1
5   0.061807      (medialuna)       1
6   0.142631            (tea)       1
7   0.103856           (cake)       1
8   0.071844       (sandwich)       1
9   0.090016  (coffee, bread)       2
10  0.054728   (coffee, cake)       2

If we print the value_counts( ) for the length column we will see that we get back a number of itemsets that contain multiple items and which contain single items.

In [ ]:
frequent_itemsets['length'].value_counts()

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


1    9
2    2
Name: length, dtype: int64

---

**Calculate association rules**

In [ ]:
from mlxtend.frequent_patterns import association_rules

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


In [ ]:
# Compute all association rules for frequent_itemsets
rules = association_rules(frequent_itemsets, metric = 'lift', min_threshold = 1.0)

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


In [ ]:
rules.head()

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


antecedents consequents  antecedent support  consequent support   support  \
0    (coffee)      (cake)            0.478394            0.103856  0.054728   
1      (cake)    (coffee)            0.103856            0.478394  0.054728   

   confidence      lift  leverage  conviction  zhangs_metric  
0    0.114399  1.101515  0.005044    1.011905       0.176684  
1    0.526958  1.101515  0.005044    1.102664       0.102840

The {cake → coffee} rule has the highest confidence at 52.7%. However, both cake and coffee appear frequently across all transactions (see frequent_itemsets), so their association could simply be a fluke. As it turns out, lift is greater than  1.0. This does give us good confidence that the association rule we recommended did not arise by random chance.



As a result, if item X and Y are bought together more frequently, then several steps can be taken to increase the profit. For instance:

* Both X and Y can be placed on the same shelf, so that buyers of one item would be prompted to buy the other.

* Promotional discounts could be applied to just one out of the two items.

* Advertisements on X could be targeted at buyers who purchase Y.

* X and Y could be combined into a new product, such as having Y in flavors of X.